Pb error msgs

tyeth opened the pull request:
Adds extra logging on encoding errors for protobuf/nanopb. 
Aimed at better logging for better diagnosing of all connectivity / reliability issues, but specifically an issue with pyportal titanos (#585).

Also incorporates PR #589 for RSSI logging to serial (with PR feedback), and PR #590 resetting the esp32 Airlift coprocessor on disconnect which also checks for connectivity if an unsuccessful MQTT ping and triggers reconnect if necessary.

 Lastly adds a macro to "retry a function until timeout with custom retry delay, while verifying return value/func-argument against a condition". Then refactors the networking port specific code to use it in a less delay heavy manner

tyeth commented:
Feedback welcomed, just threw this together for that PyPortal Titano user.

brentru commented:
@tyeth  Have you tested this on physical HW yet and played around with it?

tyeth commented:
Mainly just ran a few sensors and components on a board or two, didn't see
anything not working

On Fri, 17 May 2024, 18:42 Brent Rubell, ***@***.***> wrote:

> @tyeth <https://github.com/tyeth> Have you tested this on physical HW yet
> and played around with it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/584#issuecomment-2118104046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTBZ4ZYEU4ET2CGUU2YOCDZCY6Q5AVCNFSM6AAAAABHNK565KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJYGEYDIMBUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tyeth commented:
Skipping this release (beta83), will be reviewed in context (PyPortal Titano) next week, then make it into 84.

tyeth commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -249,25 +287,87 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+      RETRY_FUNCTION_UNTIL_TIMEOUT(
+          []() -> int { return WiFi.status(); }, // Function call each cycle
+          int,                                   // return type
+          lastResult, // return variable (unused here)
```
**not** `(unused here)`

tyeth commented:
@brentru ready for re-review

brentru commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+    int curMajor, curMinor, curPatch;
+    int reqMajor, reqMinor, reqPatch;
+
+    sscanf(currentVersion.c_str(), "%d.%d.%d", &curMajor, &curMinor, &curPatch);
```
Add handling for if `sscanf` returns an error

brentru commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+    int reqMajor, reqMinor, reqPatch;
+
+    sscanf(currentVersion.c_str(), "%d.%d.%d", &curMajor, &curMinor, &curPatch);
+    sscanf(requiredVersion.c_str(), "%d.%d.%d", &reqMajor, &reqMinor,
```
Add handling for if `sscanf` returns an error

brentru commented on src/network_interfaces/Wippersnapper_WIFININA.h:
```diff
@@ -1,275 +1,284 @@
-  }
-};
-
+/*!
```
Same readability issue as Wippersnapper.h

brentru reviewed (changes requested):
@tyeth Overall, this PR is great. It addresses issues when the PB encoder/decoder doesn't work as we expect, airlift-specific issues.

However, it still needs some tweaking before I want to merge. Could you review the following feedback and either comment so we can have a convo about it, or add commits to address them? Thanks!

tyeth commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+          equal to the required version, False otherwise.
+  */
+  /********************************************************/
+  bool compareVersions(const String &currentVersion,
```
switched back to const char*, been retested on pyportal titano

tyeth commented:
@brentru ready for re-review, I've dealt with the formatting issue + feedback, retested on pyportal titano.
 Found the RSSI virtual issue. I'd left a breadcrumb of info in another commit message, easily found after walking away from the problem, I just needed to call the method getRSSI unqualified (without WS.) when inside Wippersnapper.cpp/base-class.
 Lastly I was trying to print the RSSI before connect when no longer storing it, so refactored to print in the check_valid_ssid network interface methods instead (where the scan happens and we can briefly get access to all RSSIs before connecting).

tyeth commented:
Tagging @brentru as ready for re-review

brentru commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -234,7 +283,7 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
   const char *_ssid;           /*!< Network SSID. */
   const char *_pass;           /*!< Network password. */
-  String _fv;                  /*!< nina-fw firmware version. */
+  const char *_fv;             /*!< nina-fw firmware version. */
```
You changed `_fv` from a String (dynamically managed) to a `const char`. I'm not seeing where `_fv` ptr is properly initialized to a value i.e: `const char* _fv = "1.0.0"`

brentru commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -249,25 +298,86 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+          int,                                   // return type
+          lastResult,                            // return variable
+          [](int status) { return status == WL_CONNECTED; }, // check
+          20000, // timeout interval (ms)
```
Make these magic numbers macros

brentru commented on src/network_interfaces/Wippersnapper_AIRLIFT.h:
```diff
@@ -249,25 +298,86 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
+          lastResult,                            // return variable
+          [](int status) { return status == WL_CONNECTED; }, // check
+          20000, // timeout interval (ms)
+          200);  // interval between retries
```
Make this a macro

brentru commented on src/Wippersnapper.h:
```diff
@@ -59,6 +59,57 @@
+            The arguments to pass to the function.
+*/
+/**************************************************************************/
+#define RETRY_FUNCTION_UNTIL_TIMEOUT(func, result_type, result_var, condition, \
```
Does this handle negative values?

tyeth commented:
@brentru ready for review
Cleaned up the feedback, added the pico cert for staging and wifi101 ignores for esp32 / 8266, and improved the wifi network logging (mentions which SSID it's just connected to).

brentru reviewed (approved):
LGTM! This is a **lot** of changes, do you want to cut a new release for it?

tyeth commented:
That's a very good idea. Easier roll back that way, but should be fine ☺️
